### PR TITLE
Return explicit reasons on bad authentication

### PR DESCRIPTION
**Pull request by fwininger**
I have one server with a time delay, but it was very complicate to understand and debug the issue in the application because ApiAuth return only false for bad authentication.

Let change that with explicit error.

### CODE REVIEW

File: lib/api_auth/base.rb [hunk 1]
```diff
@@ -33,16 +33,16 @@ def sign!(request, access_id, secret_key, options = {})
     def authentic?(request, secret_key, options = {})
       return false if secret_key.nil?
 
-      options = { :override_http_method => nil }.merge(options)
+      options = { :override_http_method => nil, :raise_authentic_errors => false }.merge(options)
```
Comment (packrat386) [inline lib/api_auth/base.rb]: maybe `:raise_authentication_errors` or one could make an `authenticate!` method.

**Comment by kjg** [comment]
Thanks so much for taking this on, having the ability to know what reason the request is inauthentic is very valuable.

My suggestion is to first just extract the checks into seperate validations that all run independantly of each other, and when each one passes or fails in can put that reason into a inauthentic_reasons hash or something along those lines. And for now the `authentic?` method can still be a true or false based on weather anything showed up in inauthentic_reasons.

Does that make sense?

I think having errors get raised for things being inauthentic can also happen, but it can and should be separate from the task of just supplying the exact reason in a new method.

**Comment by fwininger** [comment]
@kjg can you take a look ? Thanks for every comments.

**Comment by kjg** [comment]
This is great! Thanks so much! I just want to think about the names a little bit more with your help.

I like the name of the method, but maybe another name might be even more clear. Maybe something like `authenticity_report`?

I've always worried about using the term "md5_match" because we still consider a request authentic if there is no md5 provided at all. I'm not sure how to name this check using positive language though which is why I've gone with md5_mismatch in the past because lack of md5 doesn't make it a mismatch. Any further thought there?

If we're looking to possibly make request invalid if it's too far in the future too, maybe we can call that check "request_within_time_window" or something along those lines. There might be a better shorter name that I can't think of.

**Comment by fwininger** [comment]
I don't change md5_match because I prefers to have only false in authenticity_report for a failure.

**Comment by kjg** [comment]
I totally understand the desire to use only positive assertions like that. Can we come up with a better term than md5_match then, since it is a little misleading to call md5_match a "true" when no md5 was provided. Or maybe we can just leave it out of the report all together if there is no md5?

**Comment by fwininger** [comment]
I close this PR because we don't agree on the name and in the fact I don't need this anymore with https://github.com/mgomes/api_auth/pull/136.

**Comment by Spone** [comment]
I still think it's valuable to have access to an explicit error message.

**Pull request closed without merging**